src/tcp_orientation_calibration.cpp

```diff
@@ -0,0 +1,182 @@
+  {
+    ROS_INFO("Pose %d: Jog robot to a new location touching the shared position and"
+             " press enter.", count);
+    std::getline(std::cin, line); // Blocks program until enter is pressed
+    tf::StampedTransform transform;
+    try
+    {
+      listener.lookupTransform(base_frame, tool0_frame,
+                               ros::Time(0), transform);
+      Eigen::Affine3d eigen_pose;
+      tf::poseTFToEigen(transform, eigen_pose);
+      observations.push_back(eigen_pose);
+      ROS_INFO_STREAM("Pose " << count << ": captured transform:\n" << eigen_pose.matrix());
+      count++;
+    }
+    catch (const tf::TransformException& ex)
+    {
+      ROS_ERROR("%s",ex.what());
+      ros::Duration(1.0).sleep();
+      continue;
+    }
+  }
+  ROS_INFO("Calibration captured %d tool poses (out of %d requested). Computing calibration...",
+           count, num_samples);
+  tool_point_calibration::TcpCalibrationResult result =
+      tool_point_calibration::calibrateTcp(observations);
+  ROS_INFO_STREAM("Calibrated tcp (meters in xyz): [" << result.tcp_offset.transpose() << "] from " << tool0_frame);
+  ROS_INFO_STREAM("Touch point (meters in xyz): [" << result.touch_point.transpose() << "] in frame " << base_frame);
+  ROS_INFO_STREAM("Average residual: " << result.average_residual);
+  ROS_INFO_STREAM("Converged: " << result.converged);
+  // store the translation vector for the next step:
+  Vector3d tcp_transl;
+  tcp_transl = result.tcp_offset.transpose();
+  if (count < 4)
+  {
+    ROS_WARN("Computing a tool calibration w/ fewer than 4 points may produce an answer with good"
+             " convergence and residual error, but without a meaningful result. You are encouraged"
+             " to try with more points");
+  }
+  ROS_INFO("=======\n step 2: TCP Orientation calibration using three point method ========== ");
+  std::vector<Affine3d> A0n; // to store the transformation matrix for the points
+  A0n.resize( 3);
+
+  for (int i=0; i<3; i++)
+  {
+    if(i ==0)
+        ROS_INFO("First point: keep the TCP at same point where the translation calibration was done, press enter");
+    else if(i ==1)
+        ROS_INFO("Second point: move the TCP to any location along the X axis of the tool and press enter");
+    else
+        ROS_INFO("Third Point: move TCP to any location along the Y axis of the tool and press enter.");
+    std::getline(std::cin, line); // Blocks program until enter is pressed
+    tf::StampedTransform transform;
+    try
+    {
+      listener.lookupTransform(base_frame, tool0_frame, ros::Time(0), transform);
+      Eigen::Affine3d eigen_pose;
+      tf::poseTFToEigen(transform, eigen_pose); // convert transform to Affine3d
+      A0n[count] = eigen_pose;
+      ROS_INFO_STREAM("Pose " << i << ": captured transform:\n" << eigen_pose.matrix());
+    }
+    catch (const tf::TransformException& ex)
+    {
+      ROS_ERROR("%s",ex.what());
+      ros::Duration(1.0).sleep();
+      continue;
+    }
```

Comment:
If this fails there is no pose put in `A0n`, which is needed in the next step, but the code is allowed to continue.